File: mssp/examples/017-finished-is-not-complete/src/TMS/sources/full_page.py
"""A page that returns everything behind it.

It imports nothing and knows nothing about the other sources. Note what it
does NOT do: it does not declare itself complete. No unit in this example is
allowed to, and refusing that is the point of section 4.
"""
NAME = "full-page"
CAN_FAIL_WITH = ["unreadable-page"]


def collect():
    return {
        "records": [{"from": NAME, "id": f"f-{n}"} for n in range(1, 4)],
        "incomplete_because": None,
    }

File: mssp/examples/018-who-the-declaration-serves/src/TMS/sources/full_page.mjs
// Returns everything behind it, in one call, and declares nothing.
export const NAME = "full-page";
export const CAN_FAIL_WITH = ["unreadable-page"];
export const HELD = 3;

export function collect({ budget = 1 } = {}) {
  const take = Math.min(HELD, budget * 3);
  return {
    records: Array.from({ length: take }, (_, n) => ({ from: NAME, id: `f-${n + 1}` })),
    incomplete_because: take < HELD ? "budget-exhausted" : null,
  };
}
